README

			   ========================
			    Welcome to Ready Lisp!
			   ========================

The intention of this project to provide a fulling working Lisp environment
that's only a double-click away for Mac OS X users.

This package includes the following code items:

 - Aquamacs 1.6
 - SBCL (Steel Bank Common Lisp) 1.0.24
 - SLIME (Superior Lisp Interaction Mode for Emacs) CVS 2008-01-23
 - paredit.el
 - redshank.el
 - cldoc.el

It also includes the following documentation, all integrated with the Info
system (just press `C-h i' within Emacs to view them):

 - ANSI Common Lisp HyperSpec
 - SLIME documentation
 - ASDF documentation
 - SBCL user's manual

There are also HTML versions of all documentation, and PDF versions of most
documentation, available in these two locations:

  Ready Lisp.app/Contents/Resources/html
  Ready Lisp.app/Contents/Resources/pdf

Everthing has been pre-configured for you to enable all options and turn on
all modes.  If you wish to configure things, edit this file:

  Ready Lisp.app/Contents/Resources/site-lisp/init.lisp.el


Happy Lisping,
  John Wiegley @ http://www.newartisans.com


----------------------------------------------------------------------

NOTE: 64-bit and threading support are not enabled in Ready Lisp, because
neither of these is officially supported by SBCL on Mac OS X (yet).  If you
wish to enable them, you will have to build your own Ready Lisp from sources.
Fortunately, this process has been made quite simple for you:

  git clone git://github.com/jwiegley/ready-lisp.git
  cd ready-lisp
  make

This may take quite a while, but eventually it will leave a .dmg file in an
"image" sub-directory.

To build with 64-bit Intel support:

  make ARCH=x86_64

To build with threading enabled:

  make THREADING=yes

To build a universal binary, assuming you have ssh access to a PowerPC box
named "ppcbox":

  make PPC_HOST=ppcbox

If you wish to update any of the sub-packages to newer versions, edit the
Makefile to select the version(s) you want, run "git checkout <TAG>" within
sbcl to select a release tag, and finally run "make clean", followed by
another "make".


NEWS
* New in version 20090130

** Updated: paredit.el and redshank.el.

** Fixed: The core file for PowerPC machines was missing.

* New in version 20090127

** Fixed: The version number for SLIME should have been 2009-01-23, not
   2009-01-25.

** Fixed: The NEWS file was not updated for the last release.

** Fixed: The distribution image contained an extraneous disk image of
   Aquamacs.

** Fixed: When starting up, the package :SWANK-PACKAGE-FU could not be found.

** Fixed: Trying to locate symbols from SLIME resulted in an invalid path
   error.

* New in version 20090125

** Upgraded various components, the current versions are:

     Aquamacs    1.6
     SBCL        1.0.24
     SLIME       2009-01-23
     CL-FAD      0.6.2
     CL-PPCRE    2.0.1
     LOCAL-TIME  0.9.3
     SERIES      2.2.10

* New in version 20080426

** Upgraded various components, the current versions are:

     Aquamacs    1.3b
     SBCL        1.0.16
     SLIME       2008-04-24
     CL-FAD      0.6.2
     CL-PPCRE    1.3.2
     LOCAL-TIME  0.9.3
     SERIES      2.2.9

** The Intel side of Ready Lisp now runs 32-bit only, instead of 64-bit and
   32-bit as before.  This is because 64-bit support is not official yet, and
   often yields inferior stack traces to the 32-bit version.

** Threading support has been disabled.  This is a feature that is also not
   officially supported on OS X yet, and can cause unpredictable behavior and
   lock ups in the SBCL compiler.  My feeling is that if someone needs to do
   enough work with SBCL to use threads, they will also be able to build their
   own Ready Lisp (see the next note).  For most users this package, however,
   their interests are better served by sticking to what is fully supported.

** Ready Lisp is now built using a completely different Makefile system which
   makes it possible for anyone to easily create their own Ready Lisp
   distribution.  See the bottom of the README for details.

   If you have a PowerPC machine accessible via ssh, you can even build a
   fully universal build of Ready Lisp.  Here's how I build locally, where my
   PowerBook G4 is accessible at 192.168.2.102:

     git clone git://github.com/jwiegley/ready-lisp.git \
         && cd ready-lisp \
         && time sh build.sh 192.168.2.102

   This process takes a very long time (several hours), but in the ends yields
   a .dmg file which contains all that you need.  If you need to update the
   versions of the various sub-packages, please edit the Makefile.  Also,
   don't specify an IP address to build.sh if you don't have a PowerPC machine
   to build with.

* New in version 1.0.12-10.5.1-2

** A couple of pathname issues were discovered that led to the inability to
   load asdf-install (or use it).  These have all been corrected.

** If you now use asdf-install and choose a "system-wide" installation, the
   installed packages are dumped into your Application bundle.  However, due
   to the way that asdf-install itself works, if you move your application
   bundle to another directory there are symbolic links which will break.  So
   I recommend always installing new packages in your home directory instead,
   unless you intend never to move Ready Lisp from /Applications.

** The sources for SBCL are now included, meaning that if you use M-. (jump to
   definition) and pick a function like mapcar, it will drop you into the
   source code for SBCL's MAPCAR implementation.

* New in version 1.0.12-10.5.1

** Ready Lisp is now fully universal, and runs on the following platforms:

   - Intel 64-bit
   - Intel 32-bit
   - PowerPC 32-bit

   There is no port of SBCL to 64-bit PowerPC.  Experimental threading has
   been enabled for both Intel platforms.

** The following pieces were updated:

   - SBCL, to version 1.0.12.17
   - SLIME, to CVS version 2007-12-06

   Aquamacs remains at version 1.2a.

** Info documentation for the Common Lisp pieces is now bundled in.  Just type
   `C-h i' to read it.  Also, when editing Common Lisp files, you can type
   `C-h f' to instantly access the HyperSpec index.  In Emacs Lisp files,
   `C-h f' will get you help on Emacs Lisp functions.

   There is also HTML and PDF versions of all documentation in:

     Ready Lisp.app/Contents/Resources/html
     Ready Lisp.app/Contents/Resources/doc

** There are a few more Common Lisp libraries bundled in the core file with
   this release:

   - CL-FAD
   - LOCAL-TIME
   - SERIES
   - MEMOIZE
   - CL-PPCRE

   I find these libraries very handy, but mainly I'm including them because
   the upcoming release of my CL-Ledger accounting tool depends on them, so it
   will work for Ready Lisp users out-of-the-box.  See the "doc" subdirectory
   above for documentation on how to use these libraries (except memoize,
   which does not have separate documentation; use MEMOIZE:MEMOIZE-FUNCTION to
   mark a function as memoized).


build.sh
#!/bin/bash

TMPDIR=/tmp

local=false
if [[ "$1" == "--local" ]]; then
    local=true
    shift 1
fi

make=false
if [[ "$1" == "--make" ]]; then
    make=true
    shift 1
fi

PPC_HOST=$1

# Bootstrap a Ready Lisp disk image from complete scratch.  Just make sure all
# relevant version numbers and URLs have been updated and pushed to the Github
# repository first.

HERE=$(pwd)

if [[ $make == true && -d $TMPDIR/ready-lisp ]]; then
    echo Updating changed files in temporary build tree ...
    rsync -rlpgoDv		  \
	--exclude=.git/	  \
	--exclude='.git*'	  \
	--exclude='*.dmg'	  \
	--exclude='build.log' \
	--exclude='*.fasl'	  \
	./ $TMPDIR/ready-lisp/
else
    if [[ -d $TMPDIR/ready-lisp ]]; then
	echo Removing previous ready-lisp build ...
	rm -fr $TMPDIR/ready-lisp
    fi

    cd $TMPDIR

    if [[ $local == true ]]; then
	echo Copying source tree to $TMPDIR/ready-lisp ...
	rsync -a ~/src/ready-lisp . || exit 1
    else
	echo Cloning ready-lisp from GitHub to $TMPDIR/ready-lisp ...
	git clone git://github.com/jwiegley/ready-lisp.git || exit 1
    fi
fi

if [[ -n "$PPC_HOST" ]]; then
    cd $TMPDIR/ready-lisp && make PPC_HOST=$PPC_HOST || exit 1
else
    cd $TMPDIR/ready-lisp && make || exit 1
fi

DISK_IMAGE=ReadyLisp-$(date +%Y%m%d).dmg

echo Moving final disk image back to source tree ...
mv build/$DISK_IMAGE $HERE

echo Zipping disk image to guard against improper MIME types on download ...
cd $HERE && zip -r $DISK_IMAGE.zip $DISK_IMAGE

echo Ready Lisp build complete.

# build.sh ends here


site-lisp/site-start.el
;; Site startup file
;; loads osx_defaults and auctex defaults 

;; Author: David Reitter, david.reitter@gmail.com
;; Maintainer: David Reitter
;; Keywords: aquamacs
 
;; Last change: $Id: site-start.el,v 1.33 2007/09/29 11:51:08 davidswelt Exp $

;; This file is part of Aquamacs Emacs
;; http://www.aquamacs.org/


;; GNU Emacs is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation; either version 2, or (at your option)
;; any later version.

;; GNU Emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;; Boston, MA 02111-1307, USA.
 
;; Copyright (C) 2006, 2007: David Reitter
 


; NONE
;; one at the end of this file



;; Version information

(defvar aquamacs-version "1.2"
"A string with Aquamacs' version number.
The format of the string is undefined. 
For a reliable numerical representation, use `aquamacs-version-id'.")

(defvar aquamacs-version-id 100
"A float indicating Aquamacs' version number.
Full integers correspond to the third position of the public
version number, e.g. version 0.9.7 is represented as `97.x'.
Minor version numbers are reflected in the decimals. 
It is guaranteed that for any two Aquamacs releases A and B,
if aquamacs-version-id for B is higher than aquamacs-version-id 
for A, then B is newer than A.")

(defvar aquamacs-minor-version "a"
"Version code for minor maintenance releases.
Changes in this code are ignored during the online version check.")

;; only for Emacs.app
(when nil
 (setq load-path (cons "~/src" load-path))
 (setq load-path (cons "~/src/oneonone" load-path))
 (setq load-path (cons "~/src/macosx" load-path))
 (setq unicode-emacs 0)
)

(when ;; do not load this twice 
    (not (memq 'aquamacs-site-start features))
 
  (provide 'aquamacs-site-start)

  (require 'load-emacs-pre-plugins)

  ;; aquamacs-reload-preloaded-files should be set by a
  ;; plugin (preloaded) --> for Aquamacs developer(s) only.
  (when (and (boundp 'aquamacs-reload-preloaded-files)
	   aquamacs-reload-preloaded-files
	   (boundp 'aq-preloaded))
      (message "Ignoring preloaded files.")
      (mapc (lambda (p)
	      (setq features (delete p features)))
	    aq-preloaded))

  (require 'aquamacs)
  (aquamacs-setup)




  (require 'aquamacs-mode-defaults)
 
  (require 'load-emacs-plugins)

  ;; workaround - Emacs doesn't do it  (0.9.9b)
  (add-hook 'after-init-hook 'display-startup-echo-area-message 'append)

  (require 'init-lisp)
  )


site-lisp/init-lisp.el
;;; init-lisp.el --- Code to initialize the Ready Lisp environment

;; Copyright (C) 2007,2008  John Wiegley.  All rights reserved.

;; Author: John Wiegley <johnw@newartisans.com>
;; Keywords: lisp
;; location: http://www.newartisans.com/software/readylisp.html

;; This file is not part of GNU Emacs.

;; This module is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published
;; by the Free Software Foundation; either version 2, or (at your
;; option) any later version.

;; This module is distributed in the hope that it will be useful, but
;; WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
;; General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs; see the file COPYING.  If not, write to the
;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
;; Boston, MA 02111-1307, USA.

;;;_ * mule

(prefer-coding-system 'utf-8)
(set-terminal-coding-system 'utf-8)
(setq x-select-request-type '(UTF8_STRING COMPOUND_TEXT TEXT STRING))

;;;_ * cldoc

(autoload 'turn-on-cldoc-mode "cldoc" nil t)

(dolist (hook '(lisp-mode-hook
		slime-repl-mode-hook))
  (add-hook hook 'turn-on-cldoc-mode))

;;;_ * ansicl

(require 'info-look)

(dolist (mode '(lisp-mode slime-mode slime-repl-mode
			  inferior-slime-mode))
  (info-lookup-add-help :mode mode
			:regexp "[^][()'\" \t\n]+"
			:ignore-case t
			:doc-spec '(("(ansicl)Symbol Index" nil nil nil))))

(eval-after-load "lisp-mode"
  '(progn
     (define-key lisp-mode-map [(control ?h) ?f] 'info-lookup-symbol)))

(defadvice Info-exit (after remove-info-window activate)
  "When info mode is quit, remove the window."
  (if (> (length (window-list)) 1)
      (delete-window)))

;;;_ * emacs-lisp

(defun elisp-indent-or-complete (&optional arg)
  (interactive "p")
  (call-interactively 'lisp-indent-line)
  (unless (or (looking-back "^\\s-*")
	      (bolp)
	      (not (looking-back "[-A-Za-z0-9_*+/=<>!?]+")))
    (call-interactively 'lisp-complete-symbol)))

(eval-after-load "lisp-mode"
  '(progn
    (define-key emacs-lisp-mode-map [tab] 'elisp-indent-or-complete)))

;;;_ * lisp

(add-hook 'lisp-mode-hook 'turn-on-auto-fill)

(put 'iterate 'lisp-indent-function 1)
(put 'mapping 'lisp-indent-function 1)
(put 'producing 'lisp-indent-function 1)

(eval-after-load "speedbar"
 '(progn
   (add-to-list 'speedbar-obj-alist '("\\.lisp$" . ".fasl"))
   (speedbar-add-supported-extension ".lisp")))

;;;_ * paredit

(autoload 'paredit-mode "paredit"
  "Minor mode for pseudo-structurally editing Lisp code." t)

(dolist (hook '(emacs-lisp-mode-hook
		lisp-mode-hook
		slime-repl-mode-hook))
  (add-hook hook #'(lambda () (paredit-mode +1))))

;;;_ * redhank

(autoload 'redshank-mode "redshank"
  "Minor mode for restructuring Lisp code (i.e., refactoring)." t)

(dolist (hook '(emacs-lisp-mode-hook
		lisp-mode-hook
		slime-repl-mode-hook))
  (add-hook hook #'(lambda () (redshank-mode +1))))

;;;_ * slime

(require 'slime)

(slime-setup
 '(inferior-slime
   slime-asdf
   slime-autodoc
   slime-banner
   slime-c-p-c
   slime-editing-commands
   slime-fancy-inspector
   slime-fancy
   slime-fuzzy
   slime-highlight-edits
   slime-parse
   slime-presentation-streams
   slime-presentations
   slime-references
   slime-sbcl-exts
   slime-package-fu
   slime-fontifying-fu
   slime-mdot-fu
   slime-scratch
   slime-tramp
   slime-enclosing-context
   ;; slime-typeout-frame
   slime-xref-browser))

(defvar *slime-use-intel-x86-64* nil)

(defvar *sbcl-arch*
  (if (string= "i386" (shell-command-to-string "arch"))
      (if (and *slime-use-intel-x86-64*
	       (string-match "x86_64: 1"
			     (shell-command-to-string
			      "sysctl hw.optional.x86_64")))
	  "x86_64"
	"i386")
    "ppc"))

(defvar *ready-lisp-resources-path*
  (expand-file-name ".." (file-name-directory load-file-name)))

(defvar *sbcl-home-path*
  (expand-file-name "sbcl/" *ready-lisp-resources-path*))

(defvar *sbcl-lib-path*
  (expand-file-name (concat "sbcl/" (if (string= *sbcl-arch* "ppc")
					"powerpc"
				      *sbcl-arch*)
			    "/lib/sbcl/")
		    *ready-lisp-resources-path*))

(defvar *sbcl-source-path*
  (expand-file-name "sbcl/source/" *ready-lisp-resources-path*))

(defvar *sbcl-site-path*
  (expand-file-name "sbcl/site/" *ready-lisp-resources-path*))

(setenv "SBCL_HOME" *sbcl-lib-path*)

(defvar *slime-source-path*
  (expand-file-name "site-lisp/edit-modes/slime/" *ready-lisp-resources-path*))

(defvar *slime-contrib-path*
  (expand-file-name "site-lisp/edit-modes/slime/contrib/"
		    *ready-lisp-resources-path*))

;; jww (2008-04-25): Ok, the following really needs some explanation; even if
;; you understand it perfectly (you nimble CL hacker you), I'm going to need
;; to remember what this does a few months or years from now...
;;
;; The problem being addressed here is that Ready Lisp pre-binds libraries
;; like SLIME into the SBCL core file, in the interests of first-time load
;; speed.  Mainly because first impressions are important enough to be worth
;; this extra effort.
;;
;; But while this work greats, SBCL and SLIME can't handle it if you relocate
;; the new core to a machine that doesn't have the same environment
;; (pathnames, etc).  I built Ready Lisp in a particular place; you as the
;; user probably aren't going to run it from that same place; this will cause
;; disjunction between the baked pathnames in the core file and your system.
;; The end result of all this is that asdf-install will fail to work, pretty
;; cryptically too.
;;
;; The answer is to dynamically rewrite the pathnames that are loaded from the
;; core file!  Yes, this is a total hack, but the alternative is to load and
;; compile all the Lisp sources that Ready Lisp depends on at load time.  Not
;; only would this be slow, but the output generated would be confusing as
;; hell to someone who's not familiar with Common Lisp already.
;;
;; So I define this function, `slime-reconfig-load-path', which gets attached
;; to `slime-connected-hook' to run once SLIME is able to talk to the SBCL
;; inferior process.  This defun is wrapped in an `eval' form because we need
;; to insert the pathnames that were determined just above, such as
;; `*sbcl-source-path*', etc.
;;
;; This function simply calls `slime-eval-async' at the first opportunity, in
;; order to pass a Common Lisp form down to SBCL.  That form changes the
;; follow special variables:
;;
;;   swank-loader:*source-directory*
;;   swank::*load-path*
;;   <path translations for logical host SYS>
;;   <paths found in asdf::*defined-systems*>
;;
;; All of these variables and definitions are being changed to relocate to the
;; Ready Lisp application's Resources directory (and sub-directories beneath
;; it), no matter where the user happens to install Ready Lisp.  You can even
;; run it straight from the disk image!

(eval
 `(defun slime-reconfig-load-path ()
    (slime-eval-async
     '(cl:prog1
       cl:nil
       (cl:setf swank-loader:*source-directory*
		,(expand-file-name "site-lisp/edit-modes/slime/"
				   *ready-lisp-resources-path*)
		swank::*load-path*
		'(,(expand-file-name "site-lisp/edit-modes/slime/contrib/"
				     *ready-lisp-resources-path*))

		(sb-impl::logical-host-translations
		 (sb-impl::find-logical-host "SYS"))
		(cl:list
		 (cl:list
		  "SYS:SRC;**;*.*.*"
		  (cl:pathname ,(concat *sbcl-source-path* "/src/**/*.*")))
		 (cl:list
		  "SYS:CONTRIB;**;*.*.*"
		  (cl:pathname ,(concat *sbcl-source-path* "/contrib/**/*.*"))))

		(sb-impl::logical-host-canon-transls
		 (sb-impl::find-logical-host "SYS"))
		(cl:list
		 (cl:list
		  (cl:pathname "SYS:SRC;**;*.*.*")
		  (cl:pathname ,(concat *sbcl-source-path* "/src/**/*.*")))
		 (cl:list
		  (cl:pathname "SYS:CONTRIB;**;*.*.*")
		  (cl:pathname ,(concat *sbcl-source-path* "/contrib/**/*.*")))))
       (cl:let
	((sbcl-site-modules
	  (quote ,(nthcdr 2 (directory-files *sbcl-site-path*)))))
	(cl:maphash
	 (cl:lambda
	  (key value)
	  (cl:declare (cl:ignore key))
	  (cl:let
	   ((component
	     (cl:car
	      (cl:last (cl:pathname-directory
			(asdf::component-relative-pathname (cl:cdr value)))))))
	   (cl:setf (cl:slot-value (cl:cdr value) 'asdf::relative-pathname)
		    (cl:pathname
		     (cl:concatenate
		      'cl:string
		      (cl:cond
		       ((cl:member component sbcl-site-modules
				   :test (cl:symbol-function 'cl:string=))
			,*sbcl-site-path*)
		       ((cl:string= component "slime")
			,*slime-source-path*)
		       (t
			,*sbcl-lib-path*))
		      component "/")))))
	 asdf::*defined-systems*))))))

(add-hook 'slime-connected-hook 'slime-reconfig-load-path)

(setq slime-lisp-implementations
      `((sbcl
	 (,@(list "/usr/bin/arch" "-arch" *sbcl-arch*
		  (expand-file-name "sbcl" *sbcl-home-path*)
		  "--core"
		  (expand-file-name "sbcl.core-with-slime" *sbcl-lib-path*)))
	 :init (lambda (port-file _)
		 (format "(swank:start-server %S :coding-system \"utf-8-unix\")\n"
			 port-file))
	 :coding-system utf-8-unix)
	(cmucl ("lisp"))
	(ecl ("ecl"))
	(clisp ("clisp") :coding-system utf-8-unix)))

(setq slime-default-lisp 'sbcl)
(setq slime-complete-symbol*-fancy t)
(setq slime-complete-symbol-function 'slime-fuzzy-complete-symbol)

(add-hook 'slime-load-hook #'(lambda () (require 'slime-fancy)))
(add-hook 'inferior-lisp-mode-hook #'(lambda () (inferior-slime-mode t)))
(remove-hook 'lisp-mode-hook 'load-and-setup-slime)

(defun indent-or-complete (&optional arg)
  (interactive "p")
  (if (or (looking-back "^\\s-*") (bolp))
      (call-interactively 'lisp-indent-line)
    (call-interactively 'slime-indent-and-complete-symbol)))

(eval-after-load "lisp-mode"
  '(progn
     (define-key lisp-mode-map [tab] 'indent-or-complete)
     (define-key lisp-mode-map [(meta ?q)] 'slime-reindent-defun)))

(eval-after-load "slime"
  '(progn
     (define-key slime-mode-map [return] 'paredit-newline)
     (define-key slime-repl-mode-map [tab] 'indent-or-complete)
     (define-key inferior-slime-mode-map [(control ?c) (control ?p)]
       'slime-repl-previous-prompt)

     (define-key slime-mode-map [(control ?h) ?f] 'info-lookup-symbol)
     (define-key slime-repl-mode-map [(control ?h) ?f] 'info-lookup-symbol)
     (define-key inferior-slime-mode-map [(control ?h) ?f] 'info-lookup-symbol)))

(slime)

(provide 'init-lisp)

;; init-lisp.el ends here
